_ext/eclipse-base/src/main/java/com/diffplug/spotless/extra/eclipse/base/osgi/ResourceAccessor.java

```diff
@@ -0,0 +1,132 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.extra.eclipse.base.osgi;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+
+import org.eclipse.osgi.storage.bundlefile.BundleEntry;
+import org.eclipse.osgi.storage.bundlefile.BundleFile;
+import org.eclipse.osgi.storage.bundlefile.DirBundleFile;
+import org.eclipse.osgi.storage.bundlefile.ZipBundleFile;
+import org.eclipse.osgi.util.ManifestElement;
+import org.osgi.framework.BundleException;
+import org.osgi.framework.Constants;
+
+/**
+ * Helper to access resources.
+ * <p>
+ * Spotless Eclipse formatters using in many cases fat JARs, compressing multiple
+ * bundles within one JAR. Their resources can get overridden in a fat JAR.
+ * Hence they provided under the bundle activators path name.
+ * </p>
+ * <p>
+ * The resource overriding prevention is required in case the Spotless
+ * Eclipse formatters integrate third party JARs, not available via M2.
+ * Additionally it can be used in case existing third party Eclipse plugins
+ * are mocked by Spotless, requiring the provision of multiple customized
+ * plugin information (META-INF, plugin.xml) within one JAR.
+ * </p>
+ */
+class ResourceAccessor {
+	/**
+	 */
+	private final String fatJarResourcePath;
+	private final BundleFile bundleFile;
+
+	/** Resources are located in the SpotlessFramework JAR */
+	ResourceAccessor() throws BundleException {
+		this(ResourceAccessor.class);
+	}
+
+	/** Resources are located in the JAR of the given class
+	 * @throws BundleException */
+	ResourceAccessor(Class<?> clazz) throws BundleException {
+		String bundleObjPath = clazz.getName();
+		fatJarResourcePath = bundleObjPath.substring(0, bundleObjPath.lastIndexOf('.'));
+		try {
+			bundleFile = getBundlFile(clazz);
+		} catch (BundleException e) {
+			throw new BundleException(String.format("Failed to locate resources for bunlde '%s'.", clazz.getName()), e);
+		}
+	}
+
+	private static BundleFile getBundlFile(Class<?> clazz) throws BundleException {
+		URL objUrl = clazz.getProtectionDomain().getCodeSource().getLocation();
+		File jarOrDirectory = new File(objUrl.getPath());
+		if (!(jarOrDirectory.exists() && jarOrDirectory.canRead())) {
+			throw new BundleException(String.format("Path '%s' for '%s' is not accessible exist on local file system.", objUrl, clazz.getName()), BundleException.READ_ERROR);
+		}
+		try {
+			return jarOrDirectory.isDirectory() ? new DirBundleFile(jarOrDirectory, false) : new ZipBundleFile(jarOrDirectory, null, null, null);
+		} catch (IOException e) {
+			throw new BundleException(String.format("Cannot access bundle at '%s'.", jarOrDirectory), BundleException.READ_ERROR);
```

Comment:
You loose the original stack trace here. You should pass the causal exception into this one.

Reply:
@JLLeitschuh Thanks a lot.
Do you know whether I can check automatically for such an "error". It is not always, but in 95% of the cases, an error. So I was wandering whether FindBugs or something similar can do something for me, where I just exclude the 5% of the correct cases by an annotation.

Reply:
@fvgh I think Google's error-prone project catches this sort of error now, but I don't think the corresponding check has been released in a versioned release of error-prone yet.